examples/advanced/examples/reth_db_provider.rs

```diff
@@ -0,0 +1,237 @@
+/// tables and static files.
+#[derive(Clone, Debug)]
+pub struct RethDBProvider<P, T> {
+    inner: P,
+    db_path: PathBuf,
+    provider_factory: WrapProviderFactory,
+    _pd: PhantomData<T>,
+}
+
+impl<P, T> RethDBProvider<P, T> {
+    /// Create a new `RethDBProvider` instance.
+    pub fn new(inner: P, db_path: PathBuf) -> Self {
+        let db = open_db_read_only(&db_path, Default::default()).unwrap();
+        let chain_spec = ChainSpecBuilder::mainnet().build();
+        let static_file_provider =
+            StaticFileProvider::read_only(db_path.join("static_files"), false).unwrap();
+
+        let provider_factory =
+            ProviderFactory::new(db.into(), chain_spec.into(), static_file_provider);
+
+        Self {
+            inner,
+            db_path,
+            provider_factory: WrapProviderFactory::new(Arc::new(provider_factory)),
+            _pd: PhantomData,
+        }
+    }
+
+    const fn factory(&self) -> &WrapProviderFactory {
+        &self.provider_factory
+    }
+
+    /// Get the DB Path
+    pub fn db_path(&self) -> PathBuf {
+        self.db_path.clone()
+    }
+}
+
+/// Implement the `Provider` trait for the `RethDBProvider` struct.
+///
+/// This is where we override specific RPC methods to fetch from the reth-db.
+impl<P, T> Provider<T> for RethDBProvider<P, T>
+where
+    P: Provider<T>,
+    T: Transport + Clone,
+{
+    fn root(&self) -> &RootProvider<T> {
+        self.inner.root()
+    }
+
+    /// Override the `get_block_number` method to fetch the latest block number from the reth-db.
+    fn get_block_number(&self) -> ProviderCall<T, NoParams, U64, u64> {
+        let provider = self.factory().provider().map_err(TransportErrorKind::custom).unwrap();
+
+        let best = provider.best_block_number().map_err(TransportErrorKind::custom);
+
+        drop(provider);
+
+        ProviderCall::<T, NoParams, U64, u64>::ready(best)
+    }
+
+    /// Override the `get_transaction_count` method to fetch the transaction count of an address.
+    ///
+    /// `RpcWithBlock` uses `ProviderCall` under the hood.
+    fn get_transaction_count(&self, address: Address) -> RpcWithBlock<T, Address, U64, u64> {
+        let this = self.factory().clone();
+        RpcWithBlock::new_provider(move |block_id| {
+            let provider = this.provider_at(block_id).map_err(TransportErrorKind::custom).unwrap();
+
+            let maybe_acc =
+                provider.basic_account(address).map_err(TransportErrorKind::custom).unwrap();
+
+            let nonce = maybe_acc.map(|acc| acc.nonce).unwrap_or_default();
+
+            drop(provider);
+
+            ProviderCall::<T, ParamsWithBlock<Address>, U64, u64>::ready(Ok(nonce))
+        })
+    }
+}
+
+/// A helper type to get the appropriate DB provider from `reth_provider::ProviderFactory`.
+#[derive(Clone, Debug)]
+struct WrapProviderFactory {
+    inner: Arc<ProviderFactory<NodeTypesWithDBAdapter<EthereumNode, Arc<DatabaseEnv>>>>,
+}
```

Comment:
it's a bit tricky rn because `ProviderFactory` implements getters for state providers directly and not through trait: https://github.com/paradigmxyz/reth/blob/ba4e41110b87cccda21d588e0a59bc94ee910e80/crates/storage/provider/src/providers/database/mod.rs#L164

however, we could use `TryIntoHistoricalStateProvider` trait directly on db provider

so the resulting bound would look like `DatabaseProviderFactory<Provider: TryIntoHistoricalStateProvider> + BlockNumReader`